### PR TITLE
Implicit model implementation.

### PR DESCRIPTION
**Pull request by kubsta**
New pull request. 

JS

### CODE REVIEW

**Comment by pesekon2** [comment]
The tests are failing. However, if you check them, they are not failing because of different results, they are crashing because your code has new dependencies (e.g., `matplotlib`). Do you know how to add it into the GH workflows, or shall I do it (I just didn't feel like pushing commits to your branch without asking)?

**Comment by pesekon2** [comment]
> The tests are failing. However, if you check them, they are not failing because of different results, they are crashing because your code has new dependencies (e.g., `matplotlib`). Do you know how to add it into the GH workflows, or shall I do it (I just didn't feel like pushing commits to your branch without asking)?

Just to keep the track of the thing: Done last week in [5c33302](https://github.com/storm-fsv-cvut/smoderp2d/pull/232/commits/5c333028d3f51cc19df8b01506f139b665458a51).

**Comment by kubsta** [comment]
Conflicts resolved. 

JS

**Comment by landam** [comment]
Unfortunately new conflicts appeared. @pesekon2 Please can you take a look?

**Comment by pesekon2** [comment]
> Unfortunately new conflicts appeared. @pesekon2 Please can you take a look?

As new conflicts could appear again before getting ready to merge, I would recommend to skip this conflict-solving Sisyphus work and let the boys (@kavkapet, @jerabekjak) test this branch in the current state. If they find the results valuable, meaningful and whateverful and we decide to merge this PR, we can solve the conflicts and king-hell run for merge before new conflicts appear.

**Comment by kavkapet** [comment]
doesnt work under ArcGIS
image

**Comment by landam** [comment]
> doesnt work under ArcGIS
> image
> 

It seems that ArcGIS comes with older scipy, which version of ArcGIS Pro @kavkapet do you have?

**Comment by pesekon2** [comment]
> It seems that ArcGIS comes with older scipy, which version of ArcGIS Pro @kavkapet do you have?

It is correct. The exact same issue used to crash the GH actions. Changing scipy from 1.7.3 to 1.10.1 did fix the issue for GH actions.

**Comment by landam** [comment]
It should be possible to upgrade version of scipy. I will test it later today.

**Comment by landam** [comment]
VirtualBox_MS Windows_21_12_2023_19_44_50

**Comment by landam** [comment]
We tested implicit model implementation in QGIS on sample data (Nucice), result attached. Some outputs (eg. `cvsur_m3`) are empty. In contrary, we found some values eg. in `mrsearstr_pa`. @kubsta Is it expected? 
[tmpz7s03nys.zip](https://github.com/storm-fsv-cvut/smoderp2d/files/13751811/tmpz7s03nys.zip)

**Comment by kubsta** [comment]
The problem with NegativeWaterLevel is solved. Whole algorithm is moved from time_step_implicit.py to time_step.py

**Comment by kavkapet** [comment]
@kubsta @pesekon2 doesnt work with "extra outputs", 
image

**Comment by pesekon2** [comment]
> @kubsta @pesekon2 doesnt work with "extra outputs", image

This is just a `warning`. If it failed, please send the output from the error tab.

**Comment by kubsta** [comment]
I have done some cleaning (deleted parts of explicit code) and solved the problem with surface bil. The infiltration for outputs was not calculated correctly.

**Comment by jerabekjak** [comment]
There is a difference in cumulative rainfall between implicit and ref model. 
@kubsta please check how the actual rainfall is calculated. See picture bellow. 
image

**Comment by kubsta** [comment]
> There is a difference in cumulative rainfall between implicit and ref model. @kubsta please check how the actual rainfall is calculated. See picture bellow. image

Problem with cumulative rainfall has been solved

File: smoderp2d/time_step.py [hunk 1]
```diff
@@ -1,265 +1,354 @@
 # @package smoderp2d.time_step methods to perform
 #  time step, and to store intermediate variables
 
+from itertools import cycle
+from uu import Error
```
Comment (pesekon2) [inline smoderp2d/time_step.py:5]: Is it really necessary to import this special error? This one should be used only for encoding/decoding. I don't see this use in the new code.

File: smoderp2d/time_step.py [hunk 2]
```diff
@@ -1,265 +1,354 @@
 # @package smoderp2d.time_step methods to perform
 #  time step, and to store intermediate variables
 
+from itertools import cycle
```
Comment (pesekon2) [inline smoderp2d/time_step.py:4]: Import not used.

File: smoderp2d/time_step.py [hunk 3]
```diff
@@ -1,265 +1,354 @@
...
+from uu import Error
+from matplotlib.pylab import norm
 from smoderp2d.core.general import Globals, GridGlobals
 import smoderp2d.processes.rainfall as rain_f
-import smoderp2d.processes.infiltration as infilt
+import smoderp2d.processes.infiltration as infiltration
```
Comment (pesekon2) [inline smoderp2d/time_step.py:9]: Although I do agree with this change in general, I do not think it should be part of this PR but rather of a separate one. It just adds changes unrelated to the implicit solution.

File: utils/hydragramy.R [hunk 4]
```diff
@@ -19,10 +19,14 @@ root  <-  "H:/2_granty_projekty/2_Bezici/2022_RAGO/01_reseni_projektu/02_Smoderp
 
 
 root  <-  "jj-tests/"
+root  <-  "tests/data/output/"
```
Comment (pesekon2) [inline utils/hydragramy.R:22]: What is the purpose of this change? Why do we update the variable twice?

File: smoderp2d/time_step.py [hunk 5]
```diff
@@ -1,265 +1,354 @@
...
+import smoderp2d.processes.rill as rill
+from smoderp2d.core.surface import inflows_comp, surface_retention_impl
+from smoderp2d.core.surface import surface_retention_update
+from smoderp2d.core.surface import update_state1
```
Comment (pesekon2) [inline smoderp2d/time_step.py:13]: Not used anywhere.

File: smoderp2d/time_step.py [hunk 6]
```diff
@@ -1,265 +1,354 @@
...
+from smoderp2d.core.surface import update_state
+
+from smoderp2d.exceptions import NegativeWaterLevel
 
 import numpy as np
 import numpy.ma as ma
+import scipy as sp
+import scipy.optimize as spopt
```
Comment (pesekon2) [inline smoderp2d/time_step.py:21]: Not used anywhere.

File: smoderp2d/time_step.py [hunk 7]
```diff
@@ -1,265 +1,354 @@
...
+from smoderp2d.core.surface import compute_h_rill_pre
 
 
 # Class manages the one time step operation
-#
-#  the class also contains methods to store the important arrays to reload that
-#  if the time step is adjusted
-#
-class TimeStep:
-    """TODO."""
```
Comment (pesekon2) [inline smoderp2d/time_step.py:22]: Was it necessary to delete the TODO without replacing with a real docstring?

File: smoderp2d/io_functions/hydrographs.py [hunk 8]
```diff
@@ -230,4 +230,4 @@ def __del__(self):
 class HydrographsPass:
     def write_hydrographs_record(self, i, j, fc, courant, dt, surface,
                                  currRain, inStream=False, sep=SEP):
-        pass
+        pass
```
Comment (pesekon2) [inline smoderp2d/io_functions/hydrographs.py:233]: Probably pretty unnecessary change.

File: smoderp2d/flow_algorithm/D8.py [hunk 9]
```diff
@@ -87,3 +87,30 @@ def __directions(inflow, direction):
...
+                value = -1
+        else:        
+            try:
+                value = mat_fd[a][b]
+            except IndexError:
+                value = -1        
+        
+        if value == coco[k][2]:
+            inflow_dirs[k] = 1.0   
+    return inflow_dirs
```
Comment (pesekon2) [inline smoderp2d/flow_algorithm/D8.py:116]: As far as I can see, it is pretty similar to `__directionsInflow`. Isn't it possible to reuse the code somehow? To avoid duplicity and make use of its common features?

File: bin/start.py [hunk 10]
```diff
@@ -3,6 +3,7 @@
 
 
 def start_smoderp2d():
+    sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))
```
Comment (pesekon2) [inline bin/start.py:6]: What is the purpose of this change? How is it connected to the implicit solution?

File: smoderp2d/core/surface.py [hunk 11]
```diff
@@ -301,40 +232,76 @@ def update_state1(ht_1, hcrit, state):
 
     :return: TODO
     """
-    if ht_1 > hcrit:
-        if state == 0:
-            return 1
+    state = ma.where(ht_1 > hcrit,ma.where(state == 0, 1, state), state)
+    # if ht_1 > hcrit:
+    #     if state == 0:
+    #         return 1
```
Comment (pesekon2) [inline smoderp2d/core/surface.py:238]: Function `update_state1` is not used anywhere in the implicit solution. What is the purpose of modifying code that is not used anywhere here?

File: smoderp2d/core/surface.py [hunk 12]
```diff
@@ -393,56 +339,47 @@ def rill_runoff(dt, effect_vrst, ratio, h_rill, rillWidth, v_rill_rest,
...
-    )
+    vol_runoff_rill = ma.filled(q_rill,0)
+        
+    return vol_runoff_rill
 
 
-def surface_retention(bil, sur):
-    """TODO.
+def surface_retention_impl(h_sur, reten_old):
+    reten = reten_old
```
Comment (pesekon2) [inline smoderp2d/core/surface.py:348]: This line has no meaning as you never modify any of these in this function.

File: smoderp2d/core/surface.py [hunk 13]
```diff
@@ -393,56 +339,47 @@ def rill_runoff(dt, effect_vrst, ratio, h_rill, rillWidth, v_rill_rest,
...
+    c = GridGlobals.c
+   
+    inflow[1:r,0:c-1] += list_fd[1:r,0:c-1,0]*tot_flow[0:r-1,1:c] #NE
+    inflow[1:r,0:c] += list_fd[1:r,0:c,1]*tot_flow[0:r-1,0:c] #N
+    inflow[1:r,1:c] += list_fd[1:r,1:c,2]*tot_flow[0:r-1,0:c-1] #NW
+    inflow[0:r,1:c] += list_fd[0:r,1:c,3]*tot_flow[0:r,0:c-1] #W
+    inflow[0:r-1,1:c] += list_fd[0:r-1,1:c,4]*tot_flow[1:r,0:c-1] #SW
+    inflow[0:r-1,0:c] += list_fd[0:r-1,0:c,5]*tot_flow[1:r,0:c] #S
+    inflow[0:r-1,0:c-1] += list_fd[0:r-1,0:c-1,6]*tot_flow[1:r,1:c] #SE
+    inflow[0:r,0:c-1] += list_fd[0:r,0:c-1,7]*tot_flow[0:r,1:c] #E
```
Comment (pesekon2) [inline smoderp2d/core/surface.py:382]: Does this take into account the fact that some of these values might be masked? I have the feeling that these in-place operations modify even masked items (but I might be wrong, too...).

File: smoderp2d/processes/infiltration/__init__.py [hunk 14]
```diff
@@ -20,8 +20,8 @@ def set_combinatIndex(newCombinatIndex):
 def philip_infiltration(soil, bil):
     infiltration = combinatIndex[0][3]
     for z in combinatIndex:
-        if ma.all(bil < 0):
-            raise NegativeWaterLevel()
+        # if ma.all(bil < 0):
+        #     raise NegativeWaterLevel()
```
Comment (pesekon2) [inline smoderp2d/processes/infiltration/__init__.py:24]: What is the purpose of commenting this check? Do we not need to check this anymore?

File: smoderp2d/processes/infiltration/__init__.py [hunk 15]
```diff
@@ -59,3 +55,11 @@ def phlilip(k, s, deltaT, totalT, NoDataValue):
 
         # except ValueError:
     return infiltration
+
+def philip_implicit(k, s, deltaT, totalT, NoDataValue):
+    if k and s == NoDataValue:
+        infiltration = NoDataValue
+    else:
+        infiltration = (0.5 * ma.divide(s, ma.sqrt(totalT + deltaT)) + k) * \
+                deltaT
```
Comment (pesekon2) [inline smoderp2d/processes/infiltration/__init__.py:64]: Comparing the new code with the original philip - why do we not need to do this `ma.where` anymore? What is the explanation of these changes purpose?

**Review by pesekon2 — CHANGES_REQUESTED** [review]

File: smoderp2d/processes/infiltration/__init__.py (hunk 14 above)
Comment (kubsta) [reply smoderp2d/processes/infiltration/__init__.py:24]: It is done in other place.

File: smoderp2d/runoff.py [hunk 16]
```diff
@@ -231,166 +245,168 @@ def run(self):
...
-                # courant conditions is satisfied (time step did
-                # change) the iteration loop breaks
-                if delta_t_tmp == self.delta_t:
+                # if the iteration exceed the maximal amount of iteration
+                # last results are stored in hydrographs
+                # and error is raised
+                if not self.flow_control.max_iter_reached():
+                    self.hydrographs.write_hydrographs_record(
+                        0,
+                        0,
```
Comment (pesekon2) [inline smoderp2d/runoff.py:328]: @kubsta Why is `0` here now? In the explicit solution, it was `None`. Should we make two branches (`explicit -> None, implicit -> 0`) here or can we change it to `None` also for the implicit solution? As far as my tests go, the implicit solution works the same way with `0` and `None`.

File: smoderp2d/runoff.py [hunk 17]
```diff
@@ -208,11 +217,16 @@ def __init__(self, provider):
             ma.masked_array(
                 np.zeros((GridGlobals.r, GridGlobals.c)), mask=GridGlobals.masks
             ),
-            True
+            Globals.isStream
```
Comment (pesekon2) [inline smoderp2d/runoff.py:220]: What is the purpose of this change? It changes results of the explicit solution. Does it have some purpose or was it created by mistake?

File: smoderp2d/core/surface.py (hunk 13 above)
Comment (kubsta) [reply smoderp2d/core/surface.py:485]: I am not sure. Probably yes. We agreed that this will be solved in the future.

File: smoderp2d/time_step.py (hunk 7 above)
Comment (kubsta) [reply smoderp2d/time_step.py:22]: It wasn't. Probably it was my superdestructive editing in the begging of the work on this project.

File: smoderp2d/flow_algorithm/D8.py (hunk 9 above)
Comment (kubsta) [reply smoderp2d/flow_algorithm/D8.py:120]: It can be reused. I will try to edit it.

File: smoderp2d/runoff.py (hunk 16 above)
Comment (kubsta) [reply smoderp2d/runoff.py]: It was mistake or something from old version of SMODERP

File: smoderp2d/runoff.py (hunk 17 above)
Comment (kubsta) [reply smoderp2d/runoff.py:220]: Probably the residue of looking for a problem with saving the values in Streams. Corrected.

Comment (kubsta) [reply smoderp2d/runoff.py:220]: Corrected.

File: smoderp2d/processes/infiltration/__init__.py (hunk 15 above)
Comment (kubsta) [reply smoderp2d/processes/infiltration/__init__.py]: Corrected

File: smoderp2d/time_step.py (hunk 7 above)
Comment (pesekon2) [reply smoderp2d/time_step.py:22]: Change took back by one of us some time ago (not sure whom, not sure which commit anymore). Thanks.

File: bin/start.py (hunk 10 above)
Comment (pesekon2) [reply bin/start.py:6]: OK. As it does not break anything but fixes it on your laptop, this line will be kept in the code unless it causes some troubles to anybody (although it should probably be solved by setting `PYTHONPATH`).

**Comment by pesekon2** [comment]
@landam, @kubsta: Could you please try to run the failing test locally (`PYTHONPATH=. pytest -s tests/test_cmd.py --config config_files/test_stream_rill_mfda.ini --hidden_config config_files/.config_implicit.ini --reference_dir data/reference/test_implicit_stream_rill_mfda/`)? It passes for me at my laptop but fails here on github and after many hours of investigation, I am still not able to find out why (the same scipy and numpy version, not sure about the rest).

**Comment by pesekon2** [comment]
It seems that the issue is in `scipy.optimize`. It could reach [different results on different architectures](https://stackoverflow.com/questions/46043768/same-optimization-code-different-results-on-different-computers). I have reached the same results on my local laptop and on lincalc-02, with the exact same code that runs here on GH workflows. GH workflows apparently have some other finite precision.

**Comment by pesekon2** [comment]
Because of the unstable results of the implicit method, the consensus is to merge it without automatic tests and consider it an "experimental" method. If we find stable optimizer capable of delivering the same results hardware-independently, we will change it later. The code should undergo a cure at some code cosmetic salon later, too.
